File: LeetCode 30 Day Challenge/16-valid_parenthesis_string.py
class Solution:
    def checkValidString(self, s):
        count_min = count_max = 0

        for char in s:
            count_max = count_max - 1 if char == ')' else count_max + 1
            count_min = count_min + 1 if char == '(' else max(count_min - 1, 0)
            if count_max < 0:
                return False

        return count_min == 0

File: LeetCode 30 Day Challenge/16-validParenthesisString.js
const checkValidString = (s) => {
    let countMin = 0
    let countMax = 0

    for (let char of s) {
        countMax = char == ')' ? countMax - 1 : countMax + 1
        countMin = char == '(' ? countMin + 1 : Math.max(countMin - 1, 0)
        if (countMax < 0) return false
    }

    return countMin === 0
}